src/coherence/test.rs

```diff
@@ -226,3 +226,173 @@ fn overlapping_negative_impls() {
         }
     }
 }
+
+#[test]
+fn orphan_check() {
+    // These tests are largely adapted from the compile-fail coherence-*.rs tests from rustc
+
+    lowering_error! {
+        program {
+            extern trait Foo { }
+            extern struct Bar { }
+
+            impl Foo for Bar { }
+        } error_msg {
+            "impl for trait \"Foo\" violates the orphan rules"
+        }
+    }
+
+    lowering_error! {
+        program {
+            extern trait Foo { }
+
+            impl<T> Foo for T { }
+        } error_msg {
+            "impl for trait \"Foo\" violates the orphan rules"
+        }
+    }
+
+    lowering_error! {
+        program {
+            extern trait Foo<T> { }
+            struct Bar { }
+
+            impl<T> Foo<Bar> for T { }
+        } error_msg {
+            "impl for trait \"Foo\" violates the orphan rules"
+        }
+    }
+
+    // Test that the `Pair` type reports an error if it contains type
+    // parameters, even when they are covered by local types. This test
+    // was originally intended to test the opposite, but the rules changed
+    // with RFC 1023 and this became illegal.
+    lowering_error! {
+        program {
+            extern trait Remote { }
+            extern struct Pair<T, U> { }
+            struct Cover<T> { }
+
+            impl<T> Remote for Pair<T, Cover<T>> { }
+        } error_msg {
+            "impl for trait \"Remote\" violates the orphan rules"
+        }
+    }
+    lowering_error! {
+        program {
+            extern trait Remote { }
+            extern struct Pair<T, U> { }
+            struct Cover<T> { }
+
+            impl<T> Remote for Pair<Cover<T>, T> { }
+        } error_msg {
+            "impl for trait \"Remote\" violates the orphan rules"
+        }
+    }
+    lowering_error! {
+        program {
+            extern trait Remote { }
+            extern struct Pair<T, U> { }
+            struct Cover<T> { }
+
+            impl<T, U> Remote for Pair<Cover<T>, U> { }
+        } error_msg {
+            "impl for trait \"Remote\" violates the orphan rules"
+        }
+    }
+
+    lowering_error! {
+        program {
+            #[auto] extern trait Send { }
+            extern trait TheTrait<T> { }
+            extern struct isize { }
+            extern struct usize { }
+
+            struct TheType { }
+
+            impl TheTrait<TheType> for isize { }
```

Comment:
putting these three impls together into one test makes it hard to see which one is getting the error...maybe break them up? if needed we could make a macro that lets you "combine" a prefix with a number of suffixes, for convenience.

Reply:
or else we could change the error to report all the failed impls at once

Reply:
(in this case, I guess we expect only one failure?)